---
title: "Ribostan"
package: "Ribostan"
author: "Dermot Harnett"
output: 
 html_notebook:
  toc: yes
output: 
  BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Ribostan}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  

---
  
# Vignette RiboEM
  
# Setup
  
```{r loadinglibraries, include=FALSE, echo=FALSE, eval=T}
# library(Ribostan)
```



```{r}

#Loading annotation
#devtools::load_all('.')
#
#first, let's load up some test data
gtf <- system.file('extdata', 'gcv37.anno.chr22.gtf', package='Ribostan', 
  mustWork=TRUE)
fafile <- system.file('extdata', 'chr22.fa.gz', package='Ribostan',
  mustWork=TRUE)
#now load our annootation, filter out suspect ORFs, and find potential uORFs
chr22_anno <- load_annotation(anno_file, fafile, add_uorfs=TRUE)

#now load Ribo-seq data, reading in the transcript-space alignments
testbam <- system.file('extdata', 'chr22.bam', package='Ribostan',
  mustWork=TRUE)
rpfs <- get_readgr(testbam, chr22_anno)
#determine offsets by maximum CDS occupancy
offsets_df <- get_offsets(rpfs, chr22_anno)
#use our offsets to determine p-site locations
psites <- get_psite_gr(rpfs, offsets_df, chr22_anno)
#use our p-sites to filter the annotation 
chr22_anno <- periodicity_filter_uORFs(chr22_anno, psites, remove=TRUE)

#now, use Stan to estimate normalized p-site densities for our data
ritpms <- get_ritpms(psites, chr22_anno)
#and get these at the gene level (ignoring uORFs)
gritpms = gene_level_expr(ritpms, chr22_anno)

#get codon-level occupanies using RUST
rust_codon_occ_df <- get_codon_occs(psites, offsets_df, chr22_anno,
  n_genes=1000, method='RUST_glm')
#get predicted mean elongation rates for ORFs
orf_elong = get_orf_elong(chr22_anno, rust_codon_occ_df)
#and at the gene level, ingoring uORFs
gn_elong = gene_level_elong(orf_elong, ritpms, chr22_anno)

#Using the KL divergence, deterine offsets in a way that's not dependent on
#patterns at the start/stop
covgrs = list(sample1=rpfs)
metacodondf <- get_metacodon_profs(covgrs, chr22_anno)
kl_df<-get_kl_df(metacodondf, chr22_anno)
kl_offsets <- select_offsets(kl_df)
kl_offsets%>%select(p_offset,length=nreadlen)%>%
  readr::write_tsv('offsets_rustvar.tsv')
#
kl_div_plot <- plot_kl_dv(kl_df, kl_offsets)
#
allcodondt <- export_codon_dts(metacodondf, kl_offsets)
kable(allcodondt)

#Metaplots
#Kl plot
#Detect Periodicity
#Gene level Elong

```
